From 81df9d9160d034e71c711c9820d59a4b3460fad4 Mon Sep 17 00:00:00 2001
From: "Uwe L. Korn" <uwe.korn@quantco.com>
Date: Sat, 12 Feb 2022 22:08:16 +0100
Subject: [PATCH 06/14] bazel bin loader path

Run install_name_tool on the client binary before it is concatenated
with the zip archive to produce bazel-real.  After concatenation the
Mach-O __LINKEDIT segment no longer covers the end of the file, so
install_name_tool cannot process it.

---
 src/BUILD | 31 +++++++++++++++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/src/BUILD b/src/BUILD
index abded27..d218bdb 100644
--- a/src/BUILD
+++ b/src/BUILD
@@ -281,16 +281,42 @@ genrule(
 )
 
 [genrule(
-    name = "bazel-bin" + suffix,
+    name = "bazel-bin-install-name" + suffix,
     srcs = [
         "//src/main/cpp:client",
+    ],
+    outs = ["bazel-install-name" + suffix],
+    cmd = "rm -f $@; cp $(location //src/main/cpp:client) $@; chmod a+rwx $@; ${INSTALL_NAME_TOOL} -rpath ${PREFIX}/lib '@loader_path/../lib/' $@",
+    executable = 1,
+    output_to_bindir = 1,
+    visibility = [
+        "//:__pkg__",  # For distribution archive lockfile generation
+        "//scripts:__pkg__",  # For bash completion generation
+        "//scripts/packages:__pkg__",  # For installer generation
+        "//src/java:__subpackages__",  # For command line reference generation
+        "//src/test:__subpackages__",  # For integration tests
+        "//tools/compliance:__pkg__",  # For license gathering
+    ],
+) for suffix, jdk in [
+    ("-dev", "_jdk_allmodules"),
+    ("", "_jdk_minimal"),
+    ("_jdk_allmodules", "_jdk_allmodules"),
+    ("_jdk_minimal", "_jdk_minimal"),
+    ("_notools", "_notools"),
+    ("_nojdk", "_nojdk"),
+]]
+
+[genrule(
+    name = "bazel-bin" + suffix,
+    srcs = [
+        "bazel-bin-install-name" + suffix,
         "package-zip" + jdk,
     ],
     outs = ["bazel" + suffix],
     # In msys, a file path without .exe suffix(say foo), refers to a file with .exe
     # suffix(say foo.exe), if foo.exe exists and foo doesn't. So, on windows, we
     # need to remove bazel.exe first, so that cat to bazel won't fail.
-    cmd = "rm -f $@; cat $(location //src/main/cpp:client) $(location :package-zip" + jdk + ") > $@ && zip -qA $@ && chmod a+x $@",
+    cmd = "rm -f $@; cat $(location :bazel-bin-install-name" + suffix + ") $(location :package-zip" + jdk + ") > $@ && zip -qA $@ && chmod a+x $@",
     executable = 1,
     output_to_bindir = 1,
     visibility = [
